### PR TITLE
Hardening/0.21.0 migration scripts

### PR DESCRIPTION
**Pull request by fgalan**

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -1,8 +1,10 @@
-- Add: support for MongoDB 2.6 (MongoDB 2.4 should *not* be used any longer as inverted geo-queries will not work in that version) (Issue #415)
+- Add: support for MongoDB 2.6/3.0 (MongoDB 2.4 should *not* be used any longer as inverted geo-queries will not work in that version) (Issue #415)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Having checked that also work with 3.0 (from a functional point of view), let's adjust the entry.

NTC (just informative)

File: scripts/managedb/attrsvector2attrsobject.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    processed += 1
+
+    sys.stdout.write('- processing entity: %d/%d   \r' % (processed, total) )
+    sys.stdout.flush()
+
+    # Check that attribute field exists and is a vector (otherwise this entity is skipped)
+    old_attrs = doc[ATTRS]
+    if not isinstance(old_attrs, list):
+        #print '- %d: not vector attribute for entity %s. Skipping' % (processed, json.dumps(doc['_id']))
```
Comment (fgalan) [inline scripts/managedb/attrsvector2attrsobject.py]: Coment intentionally kept. Mabye I need this when using these scripts in some cases.

NTC (just informative)

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 3]
```diff
@@ -397,7 +397,7 @@ static bool updateAttribute(BSONObj& attrs, BSONObjBuilder* toSet, ContextAttrib
   actualUpdate = false;
 
   /* Attributes with metadata ID are stored as <attrName>__<ID> in the attributes embedded document */
-  std::string effectiveName = caP->name;
+  std::string effectiveName = dbDotEncode(caP->name);
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: At the end, given that these fucntions are highly coupled with "dot semantcis" I have included "Dot" in the name. If in the future they become more generic, name could be changed.

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 4]
```diff
@@ -2190,3 +2190,78 @@ void cprLookupByAttribute(EntityId&                          en,
...
+#define DOT_ESCAPE_CHAR '='
+
+/* ****************************************************************************
+*
+* dbDotEncode -
+*
+* Replace:
+*   . => =
+*/
+std::string dbDotEncode(std::string fromString)
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: You told me about this function could be simplified removing malloc(), etc. However, I prefer that you do the exact changes, as you this stuff better than me :). To be practical, if you don't want to change branch, recompile, etc. in your local environment, add the code in a comment here and I will copy-paste it to the .cpp file.

File: scripts/managedb/attrsvector2attrsobject.py (hunk 2 above)
Comment (kzangeli) [inline scripts/managedb/attrsvector2attrsobject.py]: I usually add a -v (verbose) CLI parameter for stuff like this (and a function vMsg).
NTC (change if you want :-))

File: scripts/managedb/fix_location_gjson.py [hunk 5]
```diff
@@ -58,18 +58,40 @@ def old_coordinates_format(doc):
 DB = sys.argv[1]
 COL = 'entities'
 
+# Warn user
+print "WARNING!!!! This script modifies your '%s' database. It is STRONGLY RECOMMENDED that you" % DB
```
Comment (kzangeli) [inline scripts/managedb/fix_location_gjson.py]: % DB at the end-of-line ... ?
Some Python magic ?

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: ok, NTC

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: ok, sounds sound :-)

File: test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+# 01. Update/APPEND E1 with attributes d.o.t=1 and normal=2
+# 02. Query E1 gets d.o.t=1 and normal=2
+# 03. Update/UPDATE E1 attribute d.o.t=10
+# 04. Query E1 gess d.o.t=10 and normal=2
```
Comment (kzangeli) [inline test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test]: gess => gets

File: test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+      ]
+    }
+  ],
+  "updateAction": "UPDATE"
+}'
+orionCurl --url /v1/updateContext --payload "$payload" --json
+echo
+echo
+
+echo "04. Query E1 gess d.o.t=10 and normal=2"
```
Comment (kzangeli) [inline test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test]: gess

File: test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+            "statusCode": {
+                "code": "200",
+                "reasonPhrase": "OK"
+            }
+        }
+    ]
+}
+
+
+04. Query E1 gess d.o.t=10 and normal=2
```
Comment (kzangeli) [inline test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test]: ... and gess

File: scripts/managedb/attrsvector2attrsobject.py (hunk 2 above)
Comment (fgalan) [inline scripts/managedb/attrsvector2attrsobject.py]: I think we have an issue to do an overall fix of than in all the scripts (I don't remember the number...). However, this script is so simple and for using just one time (when going from 0.20.0 to 0.21.0) that probably doesn't worth the effort.

File: scripts/managedb/fix_location_gjson.py (hunk 5 above)
Comment (fgalan) [inline scripts/managedb/fix_location_gjson.py]: The part after the final '%' are variables used to "fill the gaps" in the string. Basically, the same that printf does in C (probably this Pyton-ic thing was copyed from C :))

File: test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test]: Fixed in 969ab5c

File: test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test (hunk 7 above)
Comment (fgalan) [inline test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test]: Fixed in 969ab5c

File: test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test (hunk 8 above)
Comment (fgalan) [inline test/functionalTest/cases/878_attributes_with_dot/basic_dot_attr.test]: Fixed in 969ab5c

**Comment by fgalan** [comment]
LGTM (to @kzangeli changes in 5abbb79).

Final LGTM from @kzangeli on the rest of the PR needed.

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**